File: Rudim/Common/Zobrist.cs
using Rudim.Board;

namespace Rudim.Common
{
    public static class Zobrist
    {
        public static readonly ulong[,] ZobristTable;

        static Zobrist()
        {
            // 12 piece types (6 for each color) and 64 squares, and extra - en passant, + edge cases below
            ZobristTable = new ulong[14, 64];
            for (int piece = 0; piece < 13; piece++)
            {
                for (int square = 0; square < 64; square++)
                {
                    ZobristTable[piece, square] = Random.NextULong();
                }
            }

            // Both not needed?
            ZobristTable[13, 0] = Random.NextULong(); // White to move
            ZobristTable[13, 1] = Random.NextULong(); // Black to move

            // 16 possible castling states (4 bits)
            for (int i = 2; i < 18; i++)
            {
                ZobristTable[13, i] = Random.NextULong();
            }
        }

        public static ulong GetBoardHash(BoardState boardState)
        {
            ulong currentHash = 0;

            // Go through piece + loop Lsb() instead of squares?
            for (int square = 0; square < 64; square++)
            {
                int piece = boardState.GetPieceOn((Square)square);
                if (piece != -1)
                {
                    currentHash ^= ZobristTable[piece, square];
                }
            }

            currentHash = HashSideToMove(boardState, currentHash);

            currentHash = HashCastlingRights(boardState, currentHash);

            currentHash = HashEnPassant(boardState, currentHash);

            return currentHash;
        }

        public static ulong HashCastlingRights(BoardState boardState, ulong currentHash)
        {
            // Offset by 2 to avoid collision with side-to-move keys (which use [13, 0] and [13, 1])
            currentHash ^= ZobristTable[13, 2 + (int)boardState.Castle];
            return currentHash;
        }

        private static ulong HashSideToMove(BoardState boardState, ulong currentHash)
        {
            currentHash ^= boardState.SideToMove == Side.White ? ZobristTable[13, 0] : ZobristTable[13, 1];
            return currentHash;
        }

        public static ulong FlipSideToMoveHashes(BoardState boardState, ulong currentHash)
        {
            currentHash ^= ZobristTable[13, 0];
            currentHash ^= ZobristTable[13, 1];
            return currentHash;
        }

        public static ulong HashEnPassant(BoardState boardState, ulong currentHash)
        {
            if (boardState.EnPassantSquare != Square.NoSquare)
            {
                currentHash ^= ZobristTable[12, (int)boardState.EnPassantSquare];
            }

            return currentHash;
        }
    }
}

File: src/common/zobrist.rs
use crate::board::state::BoardState;
use crate::common::side::Side;
use crate::common::square::Square;

const fn next_u64(mut state: u64) -> (u64, u64) {
    state ^= state << 13;
    state ^= state >> 7;
    state ^= state << 17;
    (state, state)
}

const fn generate_zobrist_table() -> [[u64; 64]; 14] {
    let mut table = [[0u64; 64]; 14];
    let mut state = 1804289383u64;

    let mut i = 0;
    while i < 14 {
        let mut j = 0;
        while j < 64 {
            let (next_state, val) = next_u64(state);
            state = next_state;
            table[i][j] = val;
            j += 1;
        }
        i += 1;
    }

    table
}

// TODO: flatten
pub static ZOBRIST_TABLE: [[u64; 64]; 14] = generate_zobrist_table();

pub fn init() {}

#[inline(always)]
pub fn zobrist_table() -> &'static [[u64; 64]; 14] {
    &ZOBRIST_TABLE
}

pub fn get_board_hash(board_state: &BoardState) -> u64 {
    let mut current_hash = 0;

    for square in 0..64 {
        let piece = board_state.get_piece_on(Square::from(square));
        if piece != -1 {
            current_hash ^= zobrist_table()[piece as usize][square];
        }
    }

    current_hash = hash_side_to_move(board_state, current_hash);
    current_hash = hash_castling_rights(board_state, current_hash);
    current_hash = hash_en_passant(board_state, current_hash);

    current_hash
}

pub fn hash_castling_rights(board_state: &BoardState, current_hash: u64) -> u64 {
    // Offset by 2 to avoid collision with side-to-move keys (which use [13][0] and [13][1])
    current_hash ^ zobrist_table()[13][2 + board_state.castle.bits() as usize]
}

fn hash_side_to_move(board_state: &BoardState, current_hash: u64) -> u64 {
    current_hash
        ^ if board_state.side_to_move == Side::White {
            zobrist_table()[13][0]
        } else {
            zobrist_table()[13][1]
        }
}

pub fn flip_side_to_move_hashes(_board_state: &BoardState, current_hash: u64) -> u64 {
    current_hash ^ zobrist_table()[13][0] ^ zobrist_table()[13][1]
}

pub fn hash_en_passant(board_state: &BoardState, current_hash: u64) -> u64 {
    if board_state.en_passant_square != Square::NoSquare {
        current_hash ^ zobrist_table()[12][board_state.en_passant_square as usize]
    } else {
        current_hash
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::common::helpers::STARTING_FEN;

    #[test]
    fn test_starting_hash_is_deterministic() {
        let board = BoardState::parse_fen(STARTING_FEN);
        let hash = get_board_hash(&board);
        assert_eq!(hash, 17316932686648747093);
    }
}
